crates/worker/src/partition/leadership/self_proposer.rs

```diff
@@ -149,6 +151,44 @@ impl SelfProposer {
         Ok(commit_token)
     }
 
+    pub async fn propose_many_with_notification(
+        &mut self,
+        records: impl ExactSizeIterator<Item = IngestRecord>,
+    ) -> Result<CommitToken, Error> where {
+        let sender = self.bifrost_appender.sender();
+
+        // This is ideally should be implemented
+        // by using `sender.enqueue_many`
+        // but since we have no guarantee over the
+        // underlying channel size. a `reserve_many()` might
+        // block forever waiting for n permits that will
+        // never be available.
+        //
+        // sender
+        //     .enqueue_many(records)
+        //     .await
+        //     .map_err(|_| Error::SelfProposer)?;
+        //
+        // so instead we do this.
+
+        for mut record in records {
+            // todo: unfortunately we need to decode tha pyaload first although
+            // the appended will need to encode it eventually. Maybe if there
+            // is a way to pass the raw encoded data directly to the appender
+            let envelope = StorageCodec::decode(&mut record.record)?;
```

Comment:
This is indeed a pity. Let's do this as a follow-up. I think there isn't a lot missing because `Record` already supports carrying `Bytes` and we only need a way to create a `Record` from `(Keys, Bytes)` that we have available here. This is literally wasted work that we are doing here.